{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2fa410f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "print(\"Done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fc842b4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n",
      "\n",
      " * Reading in data.../oak/stanford/groups/smontgom/amarder/t21-proj/out/subset/data/10X_Healthy_Liver.umap.subset.cells_removed.h5ad\n",
      "\n",
      " * Reading in data.../oak/stanford/groups/smontgom/amarder/t21-proj/out/subset/data/10X_DownSyndrome_Liver.umap.subset.cells_removed.h5ad\n",
      "Data read completed.\n"
     ]
    }
   ],
   "source": [
    "headdir=\"/oak/stanford/groups/smontgom/amarder/t21-proj\"\n",
    "disease_status=\"Healthy\"\n",
    "sampletype=\"Liver\"\n",
    "suffix=\"subset\"\n",
    "suffixDirec=\"\"\n",
    "if suffix==\"subset\":\n",
    "    suffixDirec=\"subset\"\n",
    "else:\n",
    "    suffixDirec=\"full\"\n",
    "os.system(\"mkdir -p \"+headdir + \"/out/\" + suffixDirec)\n",
    "\n",
    "def return_fileName(headdir,disease_status,sampletype,suffix):\n",
    "    foutpath=\"\"\n",
    "    if suffix==\"subset\":\n",
    "        fout=\"10X_\" + disease_status + \"_\" + sampletype + \".umap.subset.cells_removed.h5ad\"\n",
    "        foutpath=headdir + \"/out/subset/data/\" + fout\n",
    "    # else:\n",
    "    #     fout=\"10X_\" + disease_status + \"_\" + sampletype + \".h5ad\"\n",
    "    #     foutpath=\"/oak/stanford/groups/smontgom/amarder/data/t21/ScanpyObjects\" + \"/\" + fout\n",
    "    return foutpath\n",
    "\n",
    "foutpath=return_fileName(headdir,\"Healthy\",sampletype,suffix)\n",
    "print(\"\\n * Reading in data...\" + foutpath)\n",
    "adata=sc.read_h5ad(foutpath)\n",
    "\n",
    "foutpath=return_fileName(headdir,\"DownSyndrome\",sampletype,suffix)\n",
    "print(\"\\n * Reading in data...\" + foutpath)\n",
    "adata2=sc.read_h5ad(foutpath)\n",
    "\n",
    "# disease_status=\"DownSyndrome\"\n",
    "\n",
    "print(\"Data read completed.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5a870427",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "add35792",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                          n_genes  patient       sample  n_genes_by_counts  \\\n",
      "AAACCCAAGGAACGCT-1-0-0       4870    15633      L15633A               4870   \n",
      "AAACCCAAGTAAACAC-1-0-0       1167    15633      L15633A               1167   \n",
      "AAACCCAAGTTACTCG-1-0-0       1047    15633      L15633A               1047   \n",
      "AAACCCACAAGTTTGC-1-0-0       3707    15633      L15633A               3707   \n",
      "AAACCCACACAGAGCA-1-0-0       2024    15633      L15633A               2024   \n",
      "...                           ...      ...          ...                ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1     4582    15532  T21 15532 A               4582   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1     6004    15532  T21 15532 A               6004   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1     4305    15532  T21 15532 A               4305   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1     3842    15532  T21 15532 A               3842   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1     4071    15532  T21 15532 A               4071   \n",
      "\n",
      "                          total_counts  total_counts_mt  pct_counts_mt  \\\n",
      "AAACCCAAGGAACGCT-1-0-0         26948.0            615.0       2.282173   \n",
      "AAACCCAAGTAAACAC-1-0-0          2420.0             66.0       2.727273   \n",
      "AAACCCAAGTTACTCG-1-0-0          2154.0             87.0       4.038997   \n",
      "AAACCCACAAGTTTGC-1-0-0         18166.0            627.0       3.451503   \n",
      "AAACCCACACAGAGCA-1-0-0          5762.0            256.0       4.442902   \n",
      "...                                ...              ...            ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1       28889.0            987.0       3.416525   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1       51303.0           1984.0       3.867220   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1       31588.0            227.0       0.718627   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1       18017.0            683.0       3.790864   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1       21156.0            346.0       1.635470   \n",
      "\n",
      "                          total_counts_rb  pct_counts_rb  organ     age  \\\n",
      "AAACCCAAGGAACGCT-1-0-0             9433.0      35.004452  Liver  14 pcw   \n",
      "AAACCCAAGTAAACAC-1-0-0              443.0      18.305784  Liver  14 pcw   \n",
      "AAACCCAAGTTACTCG-1-0-0              490.0      22.748375  Liver  14 pcw   \n",
      "AAACCCACAAGTTTGC-1-0-0             5833.0      32.109436  Liver  14 pcw   \n",
      "AAACCCACACAGAGCA-1-0-0             1617.0      28.063171  Liver  14 pcw   \n",
      "...                                   ...            ...    ...     ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1          13159.0      45.550209  Liver  14 pcw   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1          17883.0      34.857609  Liver  14 pcw   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1          15592.0      49.360516  Liver  14 pcw   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1           8081.0      44.852085  Liver  14 pcw   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1           8102.0      38.296463  Liver  14 pcw   \n",
      "\n",
      "                             sorting  mean reads/cell  median genes/cell  \\\n",
      "AAACCCAAGGAACGCT-1-0-0    CD34+/Lin-            31368               1993   \n",
      "AAACCCAAGTAAACAC-1-0-0    CD34+/Lin-            31368               1993   \n",
      "AAACCCAAGTTACTCG-1-0-0    CD34+/Lin-            31368               1993   \n",
      "AAACCCACAAGTTTGC-1-0-0    CD34+/Lin-            31368               1993   \n",
      "AAACCCACACAGAGCA-1-0-0    CD34+/Lin-            31368               1993   \n",
      "...                              ...              ...                ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1       CD45+            58444               2031   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1       CD45+            58444               2031   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1       CD45+            58444               2031   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1       CD45+            58444               2031   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1       CD45+            58444               2031   \n",
      "\n",
      "                            environment  # isolated cells  # estimated cells  \\\n",
      "AAACCCAAGGAACGCT-1-0-0          Healthy              8000              13177   \n",
      "AAACCCAAGTAAACAC-1-0-0          Healthy              8000              13177   \n",
      "AAACCCAAGTTACTCG-1-0-0          Healthy              8000              13177   \n",
      "AAACCCACAAGTTTGC-1-0-0          Healthy              8000              13177   \n",
      "AAACCCACACAGAGCA-1-0-0          Healthy              8000              13177   \n",
      "...                                 ...               ...                ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1  Down Syndrome             12500               9753   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1  Down Syndrome             12500               9753   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1  Down Syndrome             12500               9753   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1  Down Syndrome             12500               9753   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1  Down Syndrome             12500               9753   \n",
      "\n",
      "                             sangerID batch     patient_sample  exp  n_counts  \\\n",
      "AAACCCAAGGAACGCT-1-0-0        L15633A     0      15633 L15633A  10X   26948.0   \n",
      "AAACCCAAGTAAACAC-1-0-0        L15633A     0      15633 L15633A  10X    2420.0   \n",
      "AAACCCAAGTTACTCG-1-0-0        L15633A     0      15633 L15633A  10X    2154.0   \n",
      "AAACCCACAAGTTTGC-1-0-0        L15633A     0      15633 L15633A  10X   18166.0   \n",
      "AAACCCACACAGAGCA-1-0-0        L15633A     0      15633 L15633A  10X    5762.0   \n",
      "...                               ...   ...                ...  ...       ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1  T21 15532 A     1  15532 T21 15532 A  10X   28889.0   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1  T21 15532 A     1  15532 T21 15532 A  10X   51303.0   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1  T21 15532 A     1  15532 T21 15532 A  10X   31588.0   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1  T21 15532 A     1  15532 T21 15532 A  10X   18017.0   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1  T21 15532 A     1  15532 T21 15532 A  10X   21156.0   \n",
      "\n",
      "                          doublet_scores   S_score  G2M_score phase  PHASE  \\\n",
      "AAACCCAAGGAACGCT-1-0-0          0.074761  1.295283  -1.863093     S  G2M+S   \n",
      "AAACCCAAGTAAACAC-1-0-0          0.001761 -0.113151  -0.216919    G1     G1   \n",
      "AAACCCAAGTTACTCG-1-0-0          0.004923 -0.214953  -0.213232    G1     G1   \n",
      "AAACCCACAAGTTTGC-1-0-0          0.033905 -1.414664  -2.365484    G1     G1   \n",
      "AAACCCACACAGAGCA-1-0-0          0.057442 -0.431019  -0.570048    G1     G1   \n",
      "...                                  ...       ...        ...   ...    ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1        0.016441 -1.165551  -1.229567    G1     G1   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1        0.051443  0.273066  -0.428886     S  G2M+S   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1        0.110927 -0.167411  -1.620393    G1     G1   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1        0.011036 -0.856027  -0.989583    G1     G1   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1        0.033261  0.683036  -0.786458     S  G2M+S   \n",
      "\n",
      "                         leiden            leiden_v1            leiden_v2  \\\n",
      "AAACCCAAGGAACGCT-1-0-0       11        Progenitors,0                MEMPs   \n",
      "AAACCCAAGTAAACAC-1-0-0        4            B cells,0              B cells   \n",
      "AAACCCAAGTTACTCG-1-0-0        6             NK cells             NK cells   \n",
      "AAACCCACAAGTTTGC-1-0-0        7                 cDC2                 cDC2   \n",
      "AAACCCACACAGAGCA-1-0-0       12  Monocyte precursors  Monocyte precursors   \n",
      "...                         ...                  ...                  ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1      9        Progenitors 2      Progenitors 2,3   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1      6      Erythroid cells    Erythroid cells,0   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1     21                   21                 21,3   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1      1                 HSCs               HSCs,0   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1     21                   21                 21,0   \n",
      "\n",
      "                                      leiden_v3              leiden_v4  \\\n",
      "AAACCCAAGGAACGCT-1-0-0                    MEMPs                  MEMPs   \n",
      "AAACCCAAGTAAACAC-1-0-0                  B cells                B cells   \n",
      "AAACCCAAGTTACTCG-1-0-0                 NK cells               NK cells   \n",
      "AAACCCACAAGTTTGC-1-0-0                     cDC2                   cDC2   \n",
      "AAACCCACACAGAGCA-1-0-0      Monocyte precursors    Monocyte precursors   \n",
      "...                                         ...                    ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1  Early Erythroid cells  Early Erythroid cells   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1   Late Erythroid cells   Late Erythroid cells   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1                B cells                B cells   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1              HSCs/MPPs              HSCs/MPPs   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1                B cells                B cells   \n",
      "\n",
      "                                      leiden_v5         leiden_v5_split  \\\n",
      "AAACCCAAGGAACGCT-1-0-0               Mast cells              Mast cells   \n",
      "AAACCCAAGTAAACAC-1-0-0                  B cells                 B cells   \n",
      "AAACCCAAGTTACTCG-1-0-0                 NK cells                NK cells   \n",
      "AAACCCACAAGTTTGC-1-0-0                     cDC2                  cDC2,4   \n",
      "AAACCCACACAGAGCA-1-0-0     Monocyte progenitors  Monocyte progenitors,0   \n",
      "...                                         ...                     ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1  Early Erythroid cells                     NaN   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1   Late Erythroid cells                     NaN   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1                B cells                     NaN   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1              HSCs/MPPs                     NaN   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1                B cells                     NaN   \n",
      "\n",
      "                                      leiden_v6         leiden_v7_split  \\\n",
      "AAACCCAAGGAACGCT-1-0-0               Mast cells            Mast cells,1   \n",
      "AAACCCAAGTAAACAC-1-0-0                  B cells                 B cells   \n",
      "AAACCCAAGTTACTCG-1-0-0                 NK cells                NK cells   \n",
      "AAACCCACAAGTTTGC-1-0-0                     cDC2                    cDC2   \n",
      "AAACCCACACAGAGCA-1-0-0     Monocyte progenitors  Monocyte progenitors,4   \n",
      "...                                         ...                     ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1  Early erythroid cells                     NaN   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1   Late erythroid cells                     NaN   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1                   pDCs                     NaN   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1              HSCs/MPPs                     NaN   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1                B cells                     NaN   \n",
      "\n",
      "                                      leiden_v7       annotations_atac  \\\n",
      "AAACCCAAGGAACGCT-1-0-0    Early erythroid cells  Early erythroid cells   \n",
      "AAACCCAAGTAAACAC-1-0-0                  B cells                B cells   \n",
      "AAACCCAAGTTACTCG-1-0-0                 NK cells               NK cells   \n",
      "AAACCCACAAGTTTGC-1-0-0                     cDC2                   cDC2   \n",
      "AAACCCACACAGAGCA-1-0-0     Monocyte progenitors   Monocyte progenitors   \n",
      "...                                         ...                    ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1  Early erythroid cells  Early erythroid cells   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1   Late erythroid cells   Late erythroid cells   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1                   pDCs                   pDCs   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1              HSCs/MPPs              HSCs/MPPs   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1                B cells    Pre pro/pro B cells   \n",
      "\n",
      "                            annotations_atac_v2 cell_type_groups  \\\n",
      "AAACCCAAGGAACGCT-1-0-0    Early erythroid cells        Erythroid   \n",
      "AAACCCAAGTAAACAC-1-0-0                  B cells          B cells   \n",
      "AAACCCAAGTTACTCG-1-0-0                 NK cells       NK/T cells   \n",
      "AAACCCACAAGTTTGC-1-0-0                     cDC2          Myeloid   \n",
      "AAACCCACACAGAGCA-1-0-0     Monocyte progenitors          Myeloid   \n",
      "...                                         ...              ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1                    NaN        Erythroid   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1                    NaN        Erythroid   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1                    NaN          Myeloid   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1                    NaN  HSC/Progenitors   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1                    NaN          B cells   \n",
      "\n",
      "                         numerical_labels              leiden_v8  \\\n",
      "AAACCCAAGGAACGCT-1-0-0                  5                    NaN   \n",
      "AAACCCAAGTAAACAC-1-0-0                  1                    NaN   \n",
      "AAACCCAAGTTACTCG-1-0-0                 18                    NaN   \n",
      "AAACCCACAAGTTTGC-1-0-0                 24                    NaN   \n",
      "AAACCCACACAGAGCA-1-0-0                 17                    NaN   \n",
      "...                                   ...                    ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1                4  Early erythroid cells   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1               12   Late erythroid cells   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1               24                   pDCs   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1                6              HSCs/MPPs   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1               19                B cells   \n",
      "\n",
      "                                      leiden_v9       leiden_v10_split  \\\n",
      "AAACCCAAGGAACGCT-1-0-0                      NaN                    NaN   \n",
      "AAACCCAAGTAAACAC-1-0-0                      NaN                    NaN   \n",
      "AAACCCAAGTTACTCG-1-0-0                      NaN                    NaN   \n",
      "AAACCCACAAGTTTGC-1-0-0                      NaN                    NaN   \n",
      "AAACCCACACAGAGCA-1-0-0                      NaN                    NaN   \n",
      "...                                         ...                    ...   \n",
      "CAAGCTAGTCCATAGT-1-0-0-1  Early erythroid cells  Early erythroid cells   \n",
      "CAAGCTATCTGAGTCA-1-0-0-1   Late erythroid cells   Late erythroid cells   \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1                   pDCs                   pDCs   \n",
      "CAAGGGAAGGATGGCT-1-0-0-1              HSCs/MPPs              HSCs/MPPs   \n",
      "CAAGGGACAGGTCAAG-1-0-0-1                B cells              B cells,1   \n",
      "\n",
      "                                     leiden_v10  \n",
      "AAACCCAAGGAACGCT-1-0-0                      NaN  \n",
      "AAACCCAAGTAAACAC-1-0-0                      NaN  \n",
      "AAACCCAAGTTACTCG-1-0-0                      NaN  \n",
      "AAACCCACAAGTTTGC-1-0-0                      NaN  \n",
      "AAACCCACACAGAGCA-1-0-0                      NaN  \n",
      "...                                         ...  \n",
      "CAAGCTAGTCCATAGT-1-0-0-1  Early erythroid cells  \n",
      "CAAGCTATCTGAGTCA-1-0-0-1   Late erythroid cells  \n",
      "CAAGGGAAGCTGGCTC-1-0-0-1                   pDCs  \n",
      "CAAGGGAAGGATGGCT-1-0-0-1              HSCs/MPPs  \n",
      "CAAGGGACAGGTCAAG-1-0-0-1        Pre pro B cells  \n",
      "\n",
      "[2000 rows x 45 columns]\n"
     ]
    }
   ],
   "source": [
    "# pd.set_option('display.max_columns', None)\n",
    "# print(adata3.obs)\n",
    "# pd.set_option('display.max_columns', 20)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "b55befa7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " * Cell type column name to use: leiden_v7\n",
      "\n",
      " * Cell type column name to use: leiden_v10\n"
     ]
    }
   ],
   "source": [
    "a=[columnName[8:] for columnName in adata.obs.columns if 'leiden_v' in columnName]\n",
    "colName=\"leiden_v\" + str(max([int(x) for x in a if x.isdigit()]))\n",
    "print(\"\\n * Cell type column name to use: \" + colName)\n",
    "adata.obs[\"leiden_names\"]=adata.obs[colName]\n",
    "\n",
    "a=[columnName[8:] for columnName in adata2.obs.columns if 'leiden_v' in columnName]\n",
    "colName=\"leiden_v\" + str(max([int(x) for x in a if x.isdigit()]))\n",
    "print(\"\\n * Cell type column name to use: \" + colName)\n",
    "adata2.obs[\"leiden_names\"]=adata2.obs[colName]\n",
    "\n",
    "adata=adata[:1000,]\n",
    "adata2=adata2[:1000,]\n",
    "adata3=adata.concatenate(adata2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "471225fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "cell_types_of_interest=[\"Early erythroid cells\",\"B cells\"]\n",
    "# \"Early erythroid cells\" in cell_types_of_interest\n",
    "# cell_type for cell_type in adata3.obs[[\"leiden_names\"]] \n",
    "\n",
    "i=0\n",
    "cell_type=cell_types_of_interest[i]\n",
    "ind=np.where(adata3.obs[[\"leiden_names\"]]==cell_type)[0]\n",
    "# ind=np.where(adata3.obs[[\"leiden_names\"]].isin(cell_types_of_interest))[0]\n",
    "adata4=adata3[ind,] #in cell_types_of_interest\n",
    "# adata4.obs.columns\n",
    "#adata3[adata3.obs[[\"leiden_names\"]] in [\"Early erythroid cells\",\"B cells\"],].obs[[\"leiden_names\"]].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "43b7b987",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/amarder/anaconda3/envs/minimal_env/lib/python3.6/site-packages/anndata/_core/anndata.py:1229: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sorting' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'environment' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sangerID' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'patient_sample' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_v1' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_v2' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_v3' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_v4' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_v5' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_v5_split' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_v6' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_v7_split' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_v7' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'annotations_atac' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'annotations_atac_v2' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'numerical_labels' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_names' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_v8' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_v9' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_v10_split' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'leiden_v10' as categorical\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 462 × 24568\n",
       "    obs: 'n_genes', 'patient', 'sample', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_rb', 'pct_counts_rb', 'organ', 'age', 'sorting', 'mean reads/cell', 'median genes/cell', 'environment', '# isolated cells', '# estimated cells', 'sangerID', 'batch', 'patient_sample', 'exp', 'n_counts', 'doublet_scores', 'S_score', 'G2M_score', 'phase', 'PHASE', 'leiden', 'leiden_v1', 'leiden_v2', 'leiden_v3', 'leiden_v4', 'leiden_v5', 'leiden_v5_split', 'leiden_v6', 'leiden_v7_split', 'leiden_v7', 'annotations_atac', 'annotations_atac_v2', 'cell_type_groups', 'numerical_labels', 'leiden_names', 'leiden_v8', 'leiden_v9', 'leiden_v10_split', 'leiden_v10'\n",
       "    var: 'n_cells_L15633A', 'n_cells_L15633B', 'n_cells_L15633R', 'n_cells_F15633C', 'n_cells_F15633S', 'n_cells_L15657A', 'n_cells_F15657F', 'n_cells_L15781A', 'n_cells_L15781C', 'n_cells_L15781D', 'Ensembl', 'Chromosome', 'feature_types', 'n_cells_F15781B-0', 'n_cells_F15657G-0', 'mean-0', 'std-0', 'highly_variable-0', 'means-0', 'dispersions-0', 'dispersions_norm-0', 'highly_variable_nbatches-0', 'highly_variable_intersection-0', 'n_cells_F15781B-1', 'n_cells_F15657G-1', 'mean-1', 'std-1', 'highly_variable-1', 'means-1', 'dispersions-1', 'dispersions_norm-1', 'highly_variable_nbatches-1', 'highly_variable_intersection-1', 'n_cells_T21 15532 A-1', 'n_cells_T21 15532 B-1', 'n_cells_T21 15562 B-1', 'n_cells_T21 15582 A-1', 'n_cells_T21 15582 B-1', 'n_cells_15582E-1', 'n_cells_15582F-1', 'n_cells_15582G-1', 'n_cells_F15582A-1', 'n_cells_F15582C-1', 'n_cells_L15593E-1', 'n_cells_L15593F-1', 'n_cells_T21 15532 C-1', 'n_cells_L15593N-1', 'n_cells_F15593A-1', 'n_cells_F15593P-1', 'n_cells_15619A-1', 'n_cells_15619B-1', 'n_cells_15619C-1', 'n_cells_15619D-1', 'n_cells_15619E-1', 'n_cells_F15619F-1', 'n_cells_L15636B-1', 'n_cells_T21 15532 D-1', 'n_cells_L15636C-1', 'n_cells_L15636D-1', 'n_cells_L15636Q-1', 'n_cells_F15636R-1', 'n_cells_F15636B-1', 'n_cells_15646A-1', 'n_cells_15646B-1', 'n_cells_L15646H-1', 'n_cells_F15646J-1', 'n_cells_15656C-1', 'n_cells_T21 15532 E-1', 'n_cells_15656D-1', 'n_cells_L15656I-1', 'n_cells_F15656E-1', 'n_cells_15667E-1', 'n_cells_15667F-1', 'n_cells_L15667L-1', 'n_cells_F15667C-1', 'n_cells_F15667D-1', 'n_cells_15667M-1', 'n_cells_15669G-1', 'n_cells_L15532M-1', 'n_cells_15669H-1', 'n_cells_L15669N-1', 'n_cells_F15669F-1', 'n_cells_F15669P-1', 'n_cells_15712A-1', 'n_cells_15712B-1', 'n_cells_L15712Q-1', 'n_cells_15712D-1', 'n_cells_15724C-1', 'n_cells_15724D-1', 'n_cells_T21 15559 A-1', 'n_cells_L15724T-1', 'n_cells_F15724H-1', 'n_cells_L15734A-1', 'n_cells_L15734B-1', 'n_cells_15734U-1', 'n_cells_F15734L-1', 'n_cells_F15734M-1', 'n_cells_15734V-1', 'n_cells_L15744C-1', 'n_cells_L15744D-1', 'n_cells_T21 15559 B-1', 'n_cells_15744W-1', 'n_cells_F15744P-1', 'n_cells_F15744N-1', 'n_cells_15744Z-1', 'n_cells_T21 15559 C-1', 'n_cells_T21 15562 A-1'\n",
       "    uns: 'rank_genes_groups'\n",
       "    obsm: 'X_pca', 'X_pca_harmonize', 'X_umap'"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc.tl.rank_genes_groups(adata4, 'environment', method='wilcoxon')\n",
    "adata4."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "50048ac2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'params': {'groupby': 'environment',\n",
       "  'reference': 'rest',\n",
       "  'method': 'wilcoxon',\n",
       "  'use_raw': True,\n",
       "  'layer': None,\n",
       "  'corr_method': 'benjamini-hochberg'},\n",
       " 'names': rec.array([('RPL23A', 'MIF'), ('TMEM14C', 'COX5B'),\n",
       "            ('RPS18', 'AC004556.1'), ..., ('AC004556.1', 'RPS18'),\n",
       "            ('COX5B', 'TMEM14C'), ('MIF', 'RPL23A')],\n",
       "           dtype=[('Down Syndrome', 'O'), ('Healthy', 'O')]),\n",
       " 'scores': rec.array([( 10.1552105,  10.019408 ), (  9.512411 ,   9.493173 ),\n",
       "            (  8.938645 ,   8.8803625), ..., ( -8.8803625,  -8.938645 ),\n",
       "            ( -9.493173 ,  -9.512411 ), (-10.019408 , -10.1552105)],\n",
       "           dtype=[('Down Syndrome', '<f4'), ('Healthy', '<f4')]),\n",
       " 'pvals': rec.array([(3.14131646e-24, 1.25251160e-23),\n",
       "            (1.86293563e-21, 2.24107801e-21),\n",
       "            (3.93971670e-19, 6.66425792e-19), ...,\n",
       "            (6.66425792e-19, 3.93971670e-19),\n",
       "            (2.24107801e-21, 1.86293563e-21),\n",
       "            (1.25251160e-23, 3.14131646e-24)],\n",
       "           dtype=[('Down Syndrome', '<f8'), ('Healthy', '<f8')]),\n",
       " 'pvals_adj': rec.array([(7.71758628e-20, 1.53858525e-19),\n",
       "            (1.37647011e-17, 1.37647011e-17),\n",
       "            (1.93581920e-15, 2.47353206e-15), ...,\n",
       "            (2.47353206e-15, 1.93581920e-15),\n",
       "            (1.37647011e-17, 1.37647011e-17),\n",
       "            (1.53858525e-19, 7.71758628e-20)],\n",
       "           dtype=[('Down Syndrome', '<f8'), ('Healthy', '<f8')]),\n",
       " 'logfoldchanges': rec.array([(  0.616837  ,  1.0344721 ), (  1.3169509 ,  0.78857225),\n",
       "            (  0.5498001 , 29.052404  ), ..., (-29.052404  , -0.5498001 ),\n",
       "            ( -0.78857225, -1.3169509 ), ( -1.0344721 , -0.616837  )],\n",
       "           dtype=[('Down Syndrome', '<f4'), ('Healthy', '<f4')])}"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata4.uns[\"rank_genes_groups\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "2a49f59e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>names</th>\n",
       "      <th>scores</th>\n",
       "      <th>logfoldchanges</th>\n",
       "      <th>pvals</th>\n",
       "      <th>pvals_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RPL23A</td>\n",
       "      <td>10.155210</td>\n",
       "      <td>0.616837</td>\n",
       "      <td>3.141316e-24</td>\n",
       "      <td>7.717586e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TMEM14C</td>\n",
       "      <td>9.512411</td>\n",
       "      <td>1.316951</td>\n",
       "      <td>1.862936e-21</td>\n",
       "      <td>1.376470e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RPS18</td>\n",
       "      <td>8.938645</td>\n",
       "      <td>0.549800</td>\n",
       "      <td>3.939717e-19</td>\n",
       "      <td>1.935819e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RPL3</td>\n",
       "      <td>8.874139</td>\n",
       "      <td>0.419354</td>\n",
       "      <td>7.047674e-19</td>\n",
       "      <td>2.473532e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RPL27</td>\n",
       "      <td>8.697596</td>\n",
       "      <td>0.510954</td>\n",
       "      <td>3.389915e-18</td>\n",
       "      <td>9.253714e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24563</th>\n",
       "      <td>JUND</td>\n",
       "      <td>-8.470692</td>\n",
       "      <td>-1.413554</td>\n",
       "      <td>2.439405e-17</td>\n",
       "      <td>4.994274e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24564</th>\n",
       "      <td>HLA-DRA</td>\n",
       "      <td>-8.830003</td>\n",
       "      <td>-3.440183</td>\n",
       "      <td>1.046731e-18</td>\n",
       "      <td>3.214512e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24565</th>\n",
       "      <td>AC004556.1</td>\n",
       "      <td>-8.880363</td>\n",
       "      <td>-29.052404</td>\n",
       "      <td>6.664258e-19</td>\n",
       "      <td>2.473532e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24566</th>\n",
       "      <td>COX5B</td>\n",
       "      <td>-9.493173</td>\n",
       "      <td>-0.788572</td>\n",
       "      <td>2.241078e-21</td>\n",
       "      <td>1.376470e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24567</th>\n",
       "      <td>MIF</td>\n",
       "      <td>-10.019408</td>\n",
       "      <td>-1.034472</td>\n",
       "      <td>1.252512e-23</td>\n",
       "      <td>1.538585e-19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24568 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            names     scores  logfoldchanges         pvals     pvals_adj\n",
       "0          RPL23A  10.155210        0.616837  3.141316e-24  7.717586e-20\n",
       "1         TMEM14C   9.512411        1.316951  1.862936e-21  1.376470e-17\n",
       "2           RPS18   8.938645        0.549800  3.939717e-19  1.935819e-15\n",
       "3            RPL3   8.874139        0.419354  7.047674e-19  2.473532e-15\n",
       "4           RPL27   8.697596        0.510954  3.389915e-18  9.253714e-15\n",
       "...           ...        ...             ...           ...           ...\n",
       "24563        JUND  -8.470692       -1.413554  2.439405e-17  4.994274e-14\n",
       "24564     HLA-DRA  -8.830003       -3.440183  1.046731e-18  3.214512e-15\n",
       "24565  AC004556.1  -8.880363      -29.052404  6.664258e-19  2.473532e-15\n",
       "24566       COX5B  -9.493173       -0.788572  2.241078e-21  1.376470e-17\n",
       "24567         MIF -10.019408       -1.034472  1.252512e-23  1.538585e-19\n",
       "\n",
       "[24568 rows x 5 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc.get.rank_genes_groups_df(adata4,group='Down Syndrome')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "eb5e5fc2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>names</th>\n",
       "      <th>scores</th>\n",
       "      <th>logfoldchanges</th>\n",
       "      <th>pvals</th>\n",
       "      <th>pvals_adj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MIF</td>\n",
       "      <td>10.019408</td>\n",
       "      <td>1.034472</td>\n",
       "      <td>1.252512e-23</td>\n",
       "      <td>1.538585e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COX5B</td>\n",
       "      <td>9.493173</td>\n",
       "      <td>0.788572</td>\n",
       "      <td>2.241078e-21</td>\n",
       "      <td>1.376470e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AC004556.1</td>\n",
       "      <td>8.880363</td>\n",
       "      <td>29.052404</td>\n",
       "      <td>6.664258e-19</td>\n",
       "      <td>2.473532e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HLA-DRA</td>\n",
       "      <td>8.830003</td>\n",
       "      <td>3.440183</td>\n",
       "      <td>1.046731e-18</td>\n",
       "      <td>3.214512e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>JUND</td>\n",
       "      <td>8.470692</td>\n",
       "      <td>1.413554</td>\n",
       "      <td>2.439405e-17</td>\n",
       "      <td>4.994274e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24563</th>\n",
       "      <td>RPL27</td>\n",
       "      <td>-8.697596</td>\n",
       "      <td>-0.510954</td>\n",
       "      <td>3.389915e-18</td>\n",
       "      <td>9.253714e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24564</th>\n",
       "      <td>RPL3</td>\n",
       "      <td>-8.874139</td>\n",
       "      <td>-0.419354</td>\n",
       "      <td>7.047674e-19</td>\n",
       "      <td>2.473532e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24565</th>\n",
       "      <td>RPS18</td>\n",
       "      <td>-8.938645</td>\n",
       "      <td>-0.549800</td>\n",
       "      <td>3.939717e-19</td>\n",
       "      <td>1.935819e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24566</th>\n",
       "      <td>TMEM14C</td>\n",
       "      <td>-9.512411</td>\n",
       "      <td>-1.316951</td>\n",
       "      <td>1.862936e-21</td>\n",
       "      <td>1.376470e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24567</th>\n",
       "      <td>RPL23A</td>\n",
       "      <td>-10.155210</td>\n",
       "      <td>-0.616837</td>\n",
       "      <td>3.141316e-24</td>\n",
       "      <td>7.717586e-20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24568 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            names     scores  logfoldchanges         pvals     pvals_adj\n",
       "0             MIF  10.019408        1.034472  1.252512e-23  1.538585e-19\n",
       "1           COX5B   9.493173        0.788572  2.241078e-21  1.376470e-17\n",
       "2      AC004556.1   8.880363       29.052404  6.664258e-19  2.473532e-15\n",
       "3         HLA-DRA   8.830003        3.440183  1.046731e-18  3.214512e-15\n",
       "4            JUND   8.470692        1.413554  2.439405e-17  4.994274e-14\n",
       "...           ...        ...             ...           ...           ...\n",
       "24563       RPL27  -8.697596       -0.510954  3.389915e-18  9.253714e-15\n",
       "24564        RPL3  -8.874139       -0.419354  7.047674e-19  2.473532e-15\n",
       "24565       RPS18  -8.938645       -0.549800  3.939717e-19  1.935819e-15\n",
       "24566     TMEM14C  -9.512411       -1.316951  1.862936e-21  1.376470e-17\n",
       "24567      RPL23A -10.155210       -0.616837  3.141316e-24  7.717586e-20\n",
       "\n",
       "[24568 rows x 5 columns]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc.get.rank_genes_groups_df(adata4,group='Healthy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfa95841",
   "metadata": {},
   "outputs": [],
   "source": [
    "enr_res = gseapy.enrichr(gene_list=glist,\n",
    "                     organism='Human',\n",
    "                     gene_sets='GO_Biological_Process_2018',\n",
    "                     description='pathway',\n",
    "                     cutoff = 0.5)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:minimal_env] *",
   "language": "python",
   "name": "conda-env-minimal_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
